core/src/main/java/com/linecorp/armeria/server/ExceptionReportingServerErrorHandler.java

```diff
@@ -0,0 +1,152 @@
+
+    @Nullable
+    private static ExceptionReportingServerErrorHandler exceptionReportingServerErrorHandler;
+    private final ServerErrorHandler delegate;
+    private final Duration duration;
+    private final LongAdder counter;
+    private long lastExceptionsCount;
+    @Nullable
+    private Throwable thrownException;
+    @Nullable
+    private ScheduledFuture<?> reportingTaskFuture;
+
+    private ExceptionReportingServerErrorHandler(ServerErrorHandler serverErrorHandler, Duration duration) {
+        assert !duration.isNegative() && !duration.isZero();
+        delegate = serverErrorHandler;
+        this.duration = duration;
+        counter = new LongAdder();
+    }
+
+    static ExceptionReportingServerErrorHandler of(ServerErrorHandler serverErrorHandler, Duration duration) {
+        if (exceptionReportingServerErrorHandler != null) {
+            return exceptionReportingServerErrorHandler;
+        }
+
+        exceptionReportingServerErrorHandler = new ExceptionReportingServerErrorHandler(serverErrorHandler,
+                                                                                        duration);
+        return exceptionReportingServerErrorHandler;
+    }
+
+    /**
+     * This method increments the {@link #counter} and stores unhandled exception that occurs. If unhandled
+     * exceptions are not previously scheduled for reporting, then they are scheduled to be reported.
+     */
+    @Nullable
+    @Override
+    public HttpResponse onServiceException(ServiceRequestContext ctx, Throwable cause) {
+        final boolean isExpectedException =
+                (cause instanceof HttpStatusException || cause instanceof HttpResponseException) &&
+                cause.getCause() != null;
+
+        if (ctx.shouldReportUnhandledExceptions() && !isExpectedException) {
+            if (reportingTaskFuture == null) {
+                reportingTaskFuture = ctx.eventLoop().scheduleAtFixedRate(
+                        this::reportException, duration.toMillis(), duration.toMillis(), TimeUnit.MILLISECONDS);
+            }
+
+            if (thrownException == null) {
+                thrownException = cause;
+            }
+            counter.increment();
+        }
+
+        return delegate.onServiceException(ctx, cause);
+    }
+
+    @Nullable
+    @Override
+    public AggregatedHttpResponse renderStatus(ServiceConfig config, @Nullable RequestHeaders headers,
+                                               HttpStatus status, @Nullable String description,
+                                               @Nullable Throwable cause) {
+        return delegate.renderStatus(config, headers, status, description, cause);
+    }
+
+    @Override
+    public void serverStarting(Server server) throws Exception {}
+
+    @Override
+    public void serverStarted(Server server) throws Exception {
+        server.config().meterRegistry().more().counter("armeria.server.unhandledExceptions",
```

Comment:
Note: this is a small nuance, but if we just use a normal counter even if multiple servers use the same meter the count will be incremented from each server.

With the current setup, if server1 registers this metric I believe the LongAdder from the second server will be ignored.
I think the current behavior is fine, but just wanted to mention it.

Reply:
You're correct. Then how about keeping two instead of one? i.e.

- Micrometer counter for metric collection; and
- `LongAdder` for logging

Not super efficient but less problems.

Reply:
Yeah, keeping 2 looks great 👍